### ISSUE TITLE
SDDrive seems broken?

### ISSUE THREAD

**Issue by jmamma** [jmamma/MIDICtrl20_MegaCommand#149]
@yatli 

Snapshot restore is broken.

**Comment by jmamma** [comment]
Tried dropping midi speed to 1x. 

Get SYSEX error code 1. Globals are all messed up.

**Comment by jmamma** [comment]
Some patterns completely erased ?

**Comment by jmamma** [comment]
Just tested a complete rec/dump using C6 and the TM-1. No problems. So problem is not on the MD side.

**Comment by jmamma** [comment]
MD.setSysexRecPos()

implementation doesnt look correct. 

edit: looks okay.

**Comment by jmamma** [comment]
Just checked 2.70 . No problems with X.03

```
commit b923b202d187df5bd879bd0a49a38c268e9f412b
Author: Yatao Li <yatli@microsoft.com>
Date:   Wed Sep 2 20:43:14 2020 +0800

    super-charged refactor: unify ElektronDevice implementations
```

I'm guessing this broke sysex transmission  ^ @yatli

**Comment by jmamma** [comment]
Won't be able to release 3.0 until this one is sorted.

https://github.com/jmamma/MIDICtrl20_MegaCommand/commit/b923b202d187df5bd879bd0a49a38c268e9f412b

**Comment by jmamma** [comment]
Kit and Global transmission is okay. It's the pattern transmission that is broken.

**Comment by jmamma** [comment]
MD SYSEX verify is not showing any transmission errors. so length + checksum calc must be okay, but payload is bad.

**Comment by jmamma** [comment]
```
 91   MDPattern() : ElektronPattern() {
 92     maxSteps = 64;
 93     maxParams = 24;
 94     maxTracks = 16;
 95     maxLocks = 64;
 96 
 97     isExtraPattern = false;
 98     init();
 99   }
100 
101   /* XXX TODO extra pattern 64 */
102 
```

^^ @yatli TODO ??

**Comment by jmamma** [comment]
```
  uint8_t i = 0;
  MD.getBlockingPattern(i);

  MD.setSysexRecPos(8, i + 1); 
  delay(20);
  MD.pattern.toSysex();
```

This is test routine is working ^ so encode /decode must be okay.

Problem must be SD storage? templates perhaps?

**Comment by jmamma** [comment]
```
 21   template <class T> bool read_data_v(T *data, FatFile *filep) {
 22     auto ret = read_data(data, sizeof(T), filep);
 25     ::new(data)T;
 26     return ret;
 27   }

::new calls the constructor.

i.e isExtraPattern = false; 
```

**Comment by jmamma** [comment]
Called by the constructor!
``` 
43     /** Clear the pattern. */
 44     void init() {
 45         clearPattern();
 46     }   
```

**Comment by yatli** [comment]
What happened today 0.0

**Comment by jmamma** [comment]
more 🐛 🐞

**Comment by yatli** [comment]
`dev` branch? I'll take a look

**Comment by jmamma** [comment]
no, master.

**Comment by jmamma** [comment]
the problem is the constructor call.. in  

 21   template <class T> bool read_data_v(T *data, FatFile *filep) {
 22     auto ret = read_data(data, sizeof(T), filep);
 25     ::new(data)T;
 26     return ret;
 27   }

it indirectly calls clearPattern();

**Comment by yatli** [comment]
Only happens for MDPattern?

**Comment by jmamma** [comment]
yes

**Comment by jmamma** [comment]
Was there a vtable in previous versions of MDPattern.. if not, the schema has changed and old .snp's won't be compatible

**Comment by yatli** [comment]
clearPattern is called even twice 0.0

**Comment by yatli** [comment]
I just checked 2.70 -- the inheritance tree is intact -- MDPattern -> ElektronPattern
And the same init sequence: MDPattern ctor calls ElektronPattern ctor calls clearPattern, then MDPattern ctor calls init()

The twist is... in 2.70, MDPattern::clearPattern is not explicitly called, **and MDPattern::clearPattern is not declared as virtual, and thus base instead calls the base implementation clearPattern() {}, so the pattern isn't cleared at all 😱**

TL;DR

I fixed a bug where the derived class doesn't properly declare virtual methods, and then it backfired (

**Comment by yatli** [comment]
@jmamma dinner time. Will fix this 0.5~1.0 hrs later

**Comment by yatli** [comment]
https://github.com/jmamma/MIDICtrl20_MegaCommand/pull/150
